pygbm/splitting.py

```diff
@@ -364,45 +378,62 @@ def _find_histogram_split_subtraction(context, feature_idx,
         context.n_bins, parent_histograms[feature_idx],
         sibling_histograms[feature_idx])
 
-    return _find_best_bin_to_split_helper(context, feature_idx, histogram)
+    return _find_best_bin_to_split_helper(context, feature_idx, histogram,
+                                          n_samples)
 
 
-@njit(locals={'gradient_left': float32, 'hessian_left': float32},
+@njit(locals={'gradient_left': float32, 'hessian_left': float32,
+              'n_samples_left': uint32},
       fastmath=True)
-def _find_best_bin_to_split_helper(context, feature_idx, histogram):
+def _find_best_bin_to_split_helper(context, feature_idx, histogram, n_samples):
     """Find best bin to split on and return the corresponding SplitInfo"""
     # Allocate the structure for the best split information. It can be
     # returned as such (with a negative gain) if the min_hessian_to_split
     # condition is not satisfied. Such invalid splits are later discarded by
     # the TreeGrower.
-    best_split = SplitInfo(-1., 0, 0, 0., 0., 0., 0.)
-
+    best_split = SplitInfo(-1., 0, 0, 0., 0., 0., 0., 0, 0)
     gradient_left, hessian_left = 0., 0.
+    n_samples_left = 0
+
     for bin_idx in range(context.n_bins):
-        gradient_left += histogram[bin_idx]['sum_gradients']
+        n_samples_left += histogram[bin_idx]['count']
+        n_samples_right = n_samples - n_samples_left
+        if context.min_samples_leaf is not None:
+            if n_samples_left < context.min_samples_leaf:
+                continue
+            if n_samples_right < context.min_samples_leaf:
+                # won't get any better
+                break
+
         if context.constant_hessian:
-            hessian_left += (histogram[bin_idx]['count'] *
-                             context.constant_hessian_value)
+            hessian_left += (histogram[bin_idx]['count']
+                             * context.constant_hessian_value)
         else:
             hessian_left += histogram[bin_idx]['sum_hessians']
         if hessian_left < context.min_hessian_to_split:
             continue
         hessian_right = context.sum_hessians - hessian_left
         if hessian_right < context.min_hessian_to_split:
-            continue
+            # won't get any better
```

Comment:
Maybe add a comment to say that the loss functions are all convex and therefore the hessians are positive.